pkg/terraform/exec/plugins/ibms3presign/resource_presign.go

```diff
@@ -0,0 +1,150 @@
+package ibms3presign
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+)
+
+var storageClass = []string{
+	"standard", "vault", "cold", "flex", "smart",
+}
+
+var endpointType = []string{
+	"public", "private", "direct",
+}
+
+func resourcePresign() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceIBMCOSBucketObjectPresignCreate,
+		Read:   resourceIBMCOSBucketObjectPresignCreate,
+		Delete: resourceIBMCOSBucketObjectPresignDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+		Schema: map[string]*schema.Schema{
+			"access_key_id": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "access_key_id",
+			},
+			"secret_access_key": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "secret_access_key",
+			},
+			"bucket_name": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "COS Bucket name",
+			},
+			"key": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "COS object key",
+			},
+			"region_location": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Region Location info.",
+			},
+			"storage_class": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validateAllowedStringValue(storageClass),
+				ForceNew:     true,
+				Description:  "Storage class info",
+			},
+			"endpoint_type": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ValidateFunc: validateAllowedStringValue([]string{"public", "private"}),
+				Description:  "public, private or direct",
+				Default:      "public",
+			},
+			"presigned_url": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Optional:    true,
+				Description: "Presigned URL",
+			},
+			"expire": {
+				Type:        schema.TypeInt,
+				Optional:    true,
+				ForceNew:    true,
+				Description: "Set expire time for URL in minutes",
+				Default:     60,
+			},
+		},
+	}
+}
+
+func validateAllowedStringValue(validValues []string) schema.SchemaValidateFunc {
+	return func(v interface{}, k string) (ws []string, errors []error) {
+		input := v.(string)
+		existed := false
+		for _, s := range validValues {
+			if s == input {
+				existed = true
+				break
+			}
+		}
+		if !existed {
+			errors = append(errors, fmt.Errorf(
+				"%q must contain a value from %#v, got %q",
+				k, validValues, input))
+		}
+		return
+
+	}
+}
+
+func resourceIBMCOSBucketObjectPresignDelete(d *schema.ResourceData, m interface{}) error {
+	return nil
```

Comment:
should we add a comment that this is unimplemented because for now we're just deleting the containing bucket? it may be a todo for anyone else who may want to utilize this

Reply:
Once we generate the presign URL that's it, we hardly can reverse or undo it.. that's the reason this function contains NoP code.. the function here is to just satisfy the terraform provider schema creation.. I will add the comment with more info